README
NAME
Geo::Coder::All - Wrapper to Geocoders 
SYNOPSIS
    use Geo::Coder::All;
    #For google geocoder
    my $google_geocoder = Geo::Coder::All->new();#geocoder defaults to Geo::Coder::Google::V3
    #You can also use optional params for google api
    my $google_geocoder = Geo::Coder::All->new(key=>'GMAP_KEY',client=>'GMAP_CLIENT');

    #For Bing
    my $bing_geocoder = Geo::Coder::All->new(geocoder=>'Bing',key=>'BING_API_KEY');

    #For Ovi
    my $ovi_geocoder = Geo::Coder::All->new(geocoder=>'Ovi');

    #For OSM
    my $osm_geocoder = Geo::Coder::All->new(geocoder=>'OSM');

    #For TomTom
    my $tomtom_geocoder = Geo::Coder::All->new(geocoder=>'TomTom');

DESCRIPTION
Geo::Coder::All is wrapper for other geocoder cpan modules such as
Geo::Coder::Google,Geo::Coder::Bing,Geo::Coder::Ovi,Geo::Coder::OSM and
Geo::Coder::TomTom. Geo::Coder::All provides common geocode output
format for all geocoder.

METHODS

geocode

For Google geocoder , we can directly set the different geocoding options when calling geocode and reverse_geocode methods. (i.e If you use Geo::Coder::Go    ogle you will have to create new instance every single time you need to change geocoding options )
    #Following will return London from United Kingdom
    $rh_location = $google_geocoder->geocode({location => 'London'});
    #With geocoding options
    #Following will return London from Canada as we used country_code => ca (country_code is ISO 3166-1 )
    $rh_location = $google_geocoder->geocode({location => 'London',language=>'en',country_code=>'ca',encoding=>'utf8',sensor=>1});
    #in spanish
    $rh_location = $google_geocoder->geocode({location => 'London',language=>'es',country_code=>'ca',encoding=>'utf8',sensor=>1});
    #default encodings is set to 'utf8' you can change to other such as 'latin1'
    #You can also set DEGUB=>1 to dump raw response from the geocoder api


You cal also set GMAO_KEY and GMAP_CLIENT directly from geocode/reverse_geocode method and it will just work

reverse_geocode

For Google reverse_geocoder

    $rh_location = $google_geocoder->reverse_geocode({latlng=>'51.508515,-0.1254872',language=>'en',encoding=>'utf8',sensor=>1})
    #in spanish
    $rh_location = $google_geocoder->reverse_geocode({latlng=>'51.508515,-0.1254872',language=>'es',encoding=>'utf8',sensor=>1})


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Geo::Coder::All

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Geo-Coder-All

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Geo-Coder-All

    CPAN Ratings
        http://cpanratings.perl.org/d/Geo-Coder-All

    Search CPAN
        http://search.cpan.org/dist/Geo-Coder-All/


LICENSE AND COPYRIGHT

Copyright (C) 2014 Rohit Deshmukh

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Perl module Geo::Coder::All

0.05_01 2014-12-15
    - Deleted META.yml and META.json, so they'll be regenerated every time
      you build a release, which ensure up-to-date entries.
    - Made the SEE ALSO entries be links to the relevant modules.
    - Reformatted this file as per CPAN::Changes::Spec
	- Updated test cases to skip tests if GeoCoder module can not be loaded

0.05 2014-10-04
	- Corrected the deps

0.04 2014-10-04
	- Allowed lazy loading for geocoders.
      If you need any other supported geo coder then you will have to
      install as required.
	- Now if you wanted to use geocoder that is not supported here then
      you can specify fully qualified class name for your custom wrapper.

0.03 2014-04-12
    - Updated test cases to use regex to match address returned by geocoder	

0.02 2014-04-08
    - updated pod 

0.01 2014-03-06
    - First version of Geo::Coder::All.




lib/Geo/Coder/All.pm
package Geo::Coder::All;
use Moose;
use namespace::autoclean;
use Module::Runtime qw(require_module);

my %VALID_GEOCODER_LIST = map { $_ => 1} qw(
    Google
    OSM
    TomTom
    Ovi
    Bing
);
has 'geocoder'              => (is=>'rw',isa=>'Str',default=>'Google');
has 'key'                   => (is=>'rw',isa=>'Str',default=>'',    reader=>'get_key');
has 'langauge'              => (is=>'rw',isa=>'Str',default=>'en',  reader=>'get_language',             init_arg=>'language');
has 'google_client'         => (is=>'rw',isa=>'Str',default=>'',    reader=>'get_google_client',        init_arg=>'client');
has 'google_apiver'         => (is=>'rw',isa=>'Num',default=>3,     reader=>'get_google_apiver',        init_arg=>'apiver');
has 'google_encoding'       => (is=>'rw',isa=>'Str',default=>'utf8',reader=>'get_google_encoding',      init_arg=>'encoding');
has 'google_country_code'   => (is=>'rw',isa=>'Str',default=>'',    reader=>'get_google_country_code',  init_arg=>'country_code');
has 'google_sensor'         => (is=>'rw',isa=>'Str',default=>'',    reader=>'get_google_sensor',        init_arg=>'sensor');

has 'geocoder_engine' => (
    is  => 'rw',
    init_arg => undef,
    lazy => 1,
    isa => 'Object',
    builder => '_build_geocoder_engine',
    handles =>{
        geocode         => 'geocode_local',
        reverse_geocode => 'reverse_geocode_local'
        } 
    );

sub _build_geocoder_engine {
    my $self        = shift;
    my $geocoder    = $self->geocoder;
    
    if(!$VALID_GEOCODER_LIST{$geocoder} && $geocoder !~ /::/ ){
        $geocoder = 'Google';
        $self->geocoder('Google');
    }
    
    my $class = ($geocoder =~ /::/ ? $geocoder : 'Geo::Coder::All::'.$geocoder);
    require_module($class);
    return $class->new(); 
}

around 'geocode' => sub{
    my ($orig,$class,$rh_args) =  @_;
    return $class->$orig($class->_process_args($rh_args));
};

around 'reverse_geocode' => sub{
    my ($orig,$class,$rh_args) =  @_;
    return $class->$orig($class->_process_args($rh_args));
};
#process the args passed to create new Geo::Coder::Google
sub _process_args {
    my ($self,$rh_args) =@_;
    $rh_args->{key}         ||= $self->get_key;
    $rh_args->{language}    ||= $self->get_language;
    $rh_args->{google_apiver}= $self->get_google_apiver || $rh_args->{apiver} if($self->geocoder eq 'Gooole');
    $rh_args->{google_client}= $self->get_google_client || $rh_args->{client} if($self->geocoder eq 'Google');
    $rh_args->{google_encoding}= $self->get_google_encoding || $rh_args->{encoding} if($self->geocoder eq 'Google');
    $rh_args->{google_country_code}= $self->get_google_country_code || $rh_args->{country_code} if($self->geocoder eq 'Google');
    return $rh_args;
}

=head1 NAME

Geo::Coder::All - Geo::Coder::All

=head1 VERSION

Version 0.05_01

=cut

our $VERSION = '0.05_01';


=head1 DESCRIPTION

Geo::Coder::All is wrapper for other geocoder cpan modules such as Geo::Coder::Google,Geo::Coder::Bing,Geo::Coder::Ovi,Geo::Coder::OSM and Geo::Coder::TomTom. Geo::Coder::All provides common geocode output format for all geocoder.


=head1 SYNOPSIS

    use Geo::Coder::All;
    #For google geocoder
    my $google_geocoder = Geo::Coder::All->new();#geocoder defaults to Geo::Coder::Google::V3
    #You can also use optional params for google api
    my $google_geocoder = Geo::Coder::All->new(key=>'GMAP_KEY',client=>'GMAP_CLIENT');

    #For Bing 
    my $bing_geocoder = Geo::Coder::All->new(geocoder=>'Bing',key=>'BING_API_KEY');

    #For Ovi 
    my $ovi_geocoder = Geo::Coder::All->new(geocoder=>'Ovi');

    #For OSM 
    my $osm_geocoder = Geo::Coder::All->new(geocoder=>'OSM');

    #For TomTom 
    my $tomtom_geocoder = Geo::Coder::All->new(geocoder=>'TomTom');

    #Currently supported geocoders are 
    Geo::Coder::Google
    Geo::Coder::Bing
    Geo::Coder::TomTom
    Geo::Coder::Ovi
    Geo::Coder::OSM
    #only Geo::Coder::Google is installed by default if you need to use other then you should install them manually

    #IF you want use geocder that is not listed above then you can now specify fully qualified class wrapper name to add your own custom handling for response. Please have look at how Geo::Coder::All::Google is working.

=head1 METHODS

Geo::Coder::All offers geocode and reverse_geocode methods

=over 2

=item geocode

For Google geocoder , we can directly set the different geocoding options when calling geocode and reverse_geocode methods. i.e If you use Geo::Coder::Google you will have to create new instance every single time you need to change geocoding options

    $rh_location = $google_geocoder->geocode({location => 'London'});
    #above will return London from United Kingdom
    #With geocoding options 
    #Following will return London from Canada as we used country_code is  ca (country_code is ISO 3166-1 )
    $rh_location = $google_geocoder->geocode({location => 'London',language=>'en',country_code=>'ca',encoding=>'utf8',sensor=>1});
    #in spanish
    $rh_location = $google_geocoder->geocode({location => 'London',language=>'es',country_code=>'ca',encoding=>'utf8',sensor=>1});
    #default encodings is set to 'utf8' you can change to other such as 'latin1'
    #You can also set DEGUB=>1 to dump raw response from the geocoder api

You cal also set GMAP_KEY and GMAP_CLIENT directly from geocode/reverse_geocode method and it will just work

=item reverse_geocode

For Google reverse_geocoder 

    $rh_location = $google_geocoder->reverse_geocode({latlng=>'51.508515,-0.1254872',language=>'en',encoding=>'utf8',sensor=>1})
    #in spanish
    $rh_location = $google_geocoder->reverse_geocode({latlng=>'51.508515,-0.1254872',language=>'es',encoding=>'utf8',sensor=>1})
    
=back

=head1 SEE ALSO

L<Geo::Coder::Many>,
L<Geo::Coder::Google>,
L<Geo::Coder::Bing>,
L<Geo::Coder::Ovi>,
L<Geo::Coder::OSM> and 
L<Geo::Coder::TomTom>.

=head1 AUTHOR

Rohit Deshmukh, C<< <raigad1630 at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-geo-coder-all at rt.cpan.org>, or through
the web interface at L<https://github.com/raigad/geo-coder-all/issues>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Geo::Coder::All

=head1 ACKNOWLEDGEMENTS

Peter Sergeant, C<< <sargie@cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2014 Rohit Deshmukh.

=cut

1; # End of Geo::Coder::All
